src/App.tsx

```diff
@@ -1,24 +1,165 @@
-import React from 'react';
+import React, { useEffect, useMemo, useState } from 'react';
 import { UserWarning } from './UserWarning';
+import { Todo } from './types/Todo';
+import * as todosService from './api/todos';
+import { TodoList } from './components/TodoList';
+import { Footer } from './components/Footer';
+import { Header } from './components/Header';
+import { Errors } from './types/Errors';
+import { FilterValue } from './types/FilterValue';
+import { ErrorNotification } from './components/ErrorNotification';
 
-const USER_ID = 0;
+const USER_ID = 12037;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterValue, setFilterValue] = useState(FilterValue.All);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [errorMessage, setErrorMessage] = useState<Errors>(Errors.Null);
+  const [loadingTodoId, setLoadingTodoId] = useState<number[]>([]);
+
+  const completedTodosCounter = todos.reduce((acc, todo) => {
+    return todo.completed ? acc + 1 : acc;
+  }, 0);
+  const activeTodosCounter = todos.length - completedTodosCounter;
+
+  const handleError = (error: Errors) => {
+    setErrorMessage(error);
+    setTimeout(() => setErrorMessage(Errors.Null), 3000);
+  };
+
+  useEffect(
+    () => {
+      todosService.getTodos(USER_ID)
+        .then(setTodos)
+        .catch(() => handleError(Errors.LoadTodos));
+    }, [],
+  );
+
+  const todosToRender = useMemo(
+    () => {
+      return todos.filter(todo => {
+        return filterValue === FilterValue.All
+          || (filterValue === FilterValue.Completed
+            ? todo.completed : !todo.completed);
+      });
+    },
+    [todos, filterValue],
+  );
+
+  const addTodo = (inputValue: string) => {
+    const data = {
+      id: 0,
+      title: inputValue,
+      completed: false,
+      userId: USER_ID,
+    };
+
+    setTempTodo(data);
+
+    todosService.createTodo(data)
+      .then((createdTodo) => {
+        setTodos((currentTodos) => [...currentTodos, createdTodo]);
+      })
+      .catch(() => {
+        handleError(Errors.AddTodo);
+      })
+      .finally(() => {
+        setTempTodo(null);
+      });
+  };
+
+  const updateTodo = async (todoToUpdate: Todo) => {
+    try {
+      const updatedTodo = await todosService
+        .updateTodo(todoToUpdate.id, todoToUpdate);
+
+      setTodos(currentTodos => {
+        return currentTodos
+          .map(todo => (todo.id === todoToUpdate.id ? updatedTodo : todo));
+      });
+    } catch (e) {
+      handleError(Errors.UpdateTodo);
+    }
+  };
+
+  const deleteTodo = (todoId: number) => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    setLoadingTodoId(completedTodos.map(todo => todo.id));
+    todosService.deleteTodo(todoId)
+      .then(() => setTodos(
+        currentTodos => currentTodos.filter(todo => todo.id !== todoId),
+      ))
+      .catch(() => handleError(Errors.DeleteTodo));
```

Comment:
```suggestion
      .catch(() => handleError(Errors.DeleteTodo))
      .finally(() => {
        setLoadingTodoId(prevIds => prevIds.filter(id => id !== todoId);
      });
```
To remove the loader add finally block and remove id from the array of loading todo ids